### DISCUSSION TITLE
Weird runtime crash with BindingReducer as a parent and on a child

### DISCUSSION THREAD

**Discussion by ricardomsm** [pointfreeco/swift-composable-architecture#1877] [Q&A] [answered]
Hello 👋🏼,

We're using `TCA`, version `0.50.0`, and we're experiencing a weird runtime crash, which much to our surprise and without great understanding only happens on `Staging` and `Release` configuration, to note, in **theory**(in what regards to build settings), `Staging` is exactly equal to `DEBUG`.

We have the following parent and child reducers:
Parent
<img width="846" alt="SCR-20230127-mdx" src="https://user-images.githubusercontent.com/41112522/215135786-18a1eda5-a5b3-4752-8327-cb49a8eb1890.png">

Child
<img width="722" alt="SCR-20230127-mje" src="https://user-images.githubusercontent.com/41112522/215135904-35e59c29-0411-41e3-b8fc-69129866e839.png">

And this is the weird runtime stack trace we get when we dismiss the child one through a `viewStore.send` in which we catch in the parent and `nil` out the child:
<img width="1913" alt="SCR-20230127-mc5" src="https://user-images.githubusercontent.com/41112522/215136292-31f496f8-8b6e-465c-93f2-f6fef84ec8c5.png">
<img width="1903" alt="SCR-20230127-mci" src="https://user-images.githubusercontent.com/41112522/215136299-006e05bc-18c5-4905-a6d1-760e21eebe4a.png">
<img width="1905" alt="SCR-20230127-md0" src="https://user-images.githubusercontent.com/41112522/215136302-2387c198-ceaa-42ca-a443-4e67b3744413.png">

Through searching and trying out some possible solutions, we found this answered issue https://github.com/pointfreeco/swift-composable-architecture/discussions/1564 which gave us a hint so we tried these two solutions that solved our problem:

Declaring the case paths explicitly without the use of the `/` operator prevented the crash,
<img width="679" alt="SCR-20230127-me2" src="https://user-images.githubusercontent.com/41112522/215136905-aef24f24-34b9-4009-9f57-06bf5fa22a23.png">

As well as changing the order of our actions in the `Action` type, by moving the `binding` action to be at the bottom, the crash was avoided 🤷🏼‍♂️🥲
<img width="304" alt="SCR-20230127-mec" src="https://user-images.githubusercontent.com/41112522/215137294-6c9ea4e4-92ff-473f-8191-7c55e4bff6fc.png">

Although we "solved" this in our case, we'd like to know if anyone else has encountered a similar issue, or if there's any hint someone can give in regards to what is actually happening behind the scenes.

Thanks in advance 🙏🏼

**Comment by stephencelis** [comment, marked as answer]
Hey @ricardomsm! You mention this being an issue in 0.50.0. Does this mean the same code paths cause no problem in 0.49?

Also, is there any chance for you to whittle down a reproduction that you can share and we can use to help troubleshoot a fix?

**Reply by ricardomsm** [reply]
We tried downgrading up until `0.45.0` but the problem persisted, forgot to mention 🥲. I've been trying to put together a small sample that would have the same structure that I showed and that we're using, but I haven't been able to replicate the problem sadly 😕.

**Reply by tgrapperon** [reply]
@ricardomsm Out of curiosity, is it still crashing if you add dummy cases to your route or make a few Action/Route enums `indirect`. This resembles https://github.com/pointfreeco/swift-case-paths/issues/71

**Reply by ricardomsm** [reply]
@tgrapperon Indeed adding an indirect case to my enum causes the crash as well, although in a little example I'm trying to craft to expose the issue, I can't reproduce it yet even with the "same" structure in terms of child and parent.

**Reply by farvalhoSH** [reply]
Hi everyone, I'm a colleague of @ricardomsm and am helping him navigate these waters.
I was following @tgrapperon's suggestion and found that while adding dummy cases and/or making some actions `indirect` didn't prevent the crash, specifically making the `BindingAction` case `indirect` did. I've tested all manner of scenarios in our project and it seems to always work as long as that specific case is `indirect`.

![image](https://user-images.githubusercontent.com/116080670/215448154-65e96904-d95f-4c8b-879a-b914a5aa26f6.png)

**Reply by tgrapperon** [reply]
Ah interesting, I always assumed that having one indirect case was equivalent to mark the whole enum indirect.

**Reply by ricardomsm** [reply]
@stephencelis We managed to wip up a little demo that has the identic stack trace we're facing.
[BindingActionCrashDemo.zip](https://github.com/pointfreeco/swift-composable-architecture/files/10535105/BindingActionCrashDemo.zip)

**Reply by stephencelis** [reply]
This is a Swift bug around existentials and enum layouts:

https://github.com/pointfreeco/swift-composable-architecture/pull/1881

We can work around it in the library, luckily, but thanks for the repro and catch!
